# The Basics of moving and controlling. 

In [1]:
import sys
#the uArm application is inside of the directory 
sys.path.append('..')
from uarm.wrapper import SwiftAPI


swift = SwiftAPI(port="COM4", callback_thread_pool_size=1)

## Set the arm back to the center. 

In [2]:
swift.reset(wait=True)

## First, where is the arm right now?


In [20]:
swift.get_position()

[100.0, 100.0, 100.0]

The angle here is in `[x,y,z]`

If you were the arm, with the cables connected to your back and looking forward.
* `x` = left and right 
* `y` = the arm stretched outward or contracted inward
* `z` = up and down. 

In [18]:
swift.set_position(x=100, y=10, z=100,wait=True)

'OK'

In [8]:
swift.get_position()

[200.0, 0.0, 150.0]

In [16]:
swift.set_position(x=100, y=100, z=100,wait=True)

'OK'

In [15]:
swift.get_position()

[100.0, 100.0, 100.0]

### Setting the speed

Each use of the `speed` keyword argument sets the global speed. Not just the speed for that movement. 

100 -- really, really slow.

1000 -- slow (default). 

**Note** - There is a middle ground between 10000 and 50000 that is kind of choppy. I'd use one of these values. 

500000 -- Relatively fast. 

If you set the the value to outside of the _undocumented_ range, none of the movements will work. 

In [122]:
# change the value of speed to see it go faster. 
swift.reset(wait=True, speed=500000)

In [121]:
# without changing the speed here, it will use the speed from above. 
swift.set_position(x=100, y=100, z=100,wait=True)

'OK'

## Setting a relative position

For incremental movements, relative to where _the arm currently is_ use `relative=True`. In this case only use very small amounts. 

Note that moving the position here may actually move more than one of the servos. This is because it is moving to a specific place on an X,Y,Z plane. 

In [108]:
swift.set_position(x=10,relative=True,wait=True)

'OK'

In [127]:
#down 
swift.set_position(x=-10,relative=True,wait=True)
swift.set_position(y=-10,relative=True,wait=True)
swift.set_position(z=-10,relative=True,wait=True)
#back
swift.set_position(x=10,relative=True,wait=True)
swift.set_position(y=10,relative=True,wait=True)
swift.set_position(z=10,relative=True,wait=True)

'OK'

# Moving individual components

## by ANGLE

servo_id：
* 0: BOTTOM - "Rotates the arm on the base"
* 1: LEFT - "Stretches the arm outward"
* 2: RIGHT - 
* 3: HAND - "Gripper wrist"

Note that 3(wrist) does not print out in the `get_servo_angle` 

when setting any angle, remember the following: 
* 0 is the far left. 
* 90 in center
* 180 is far right

However you have to think about the position of the servo, not the position of the arm. For example on the bottom (base) servo, going from 90 degrees to zero moves servo counter clockwise. However you are facing the other way so the servo turns clockwise and the arm moves to the right. 

In [29]:
# Base motor
swift.set_servo_angle(0,100)

In [30]:
swift.set_servo_angle(1,0)

In [31]:
swift.set_servo_angle(2,180)

In [32]:
swift.set_servo_angle(3,0)

In [33]:
swift.get_servo_angle()

'TIMEOUT'

In [28]:
swift.get_servo_angle(servo_id=1)

0.0

In [34]:
swift.reset(wait=True, speed=500000)

## by POSITION

Parameter：
* x：X axis value, default value: None, use the X value of last movement or initial value 150.
* y：Y axis value, default value: None, use the Y value of last movement or initial value 0.
* z：Z axis value, default value: None, use the Z value of last movement or initial value 150.
* speed：speed value, default value: None, use the speed value of last movement for default speed 1000.
* relative：is/isn’t relative movement, default value: False
* wait：do/don’t wait for return, default value: False


In [14]:
swift.set_position()

### 4th motor (the wrist arm of the grabber)

* **Angle** A value between 0 and 180. 0 = far left, 180 = far right, 90 = center

In [15]:
swift.set_wrist(10)

### the gripper (if attached)

`True` = closes the gripper
`False` = opens the gripper

In [133]:
swift.set_gripper(True)

'OK'

##  QA


In [17]:
swift.get_temperature()

{'current_temperature': 0.0, 'target_temperature': 0.0}

In [18]:
swift.get_power_status()

True

In [19]:
swift.get_device_info()

{'device_type': 'SwiftPro',
 'hardware_version': '3.3.1',
 'firmware_version': '3.2.0',
 'api_version': '3.2.0',
 'device_unique': '78DB2FB28F96'}

In [20]:
swift.get_mode()

0

In [21]:
swift.get_servo_attach()

True

## Cleanup: disconnect

In [35]:
swift.disconnect()